src/Module.h

```diff
@@ -242,23 +279,107 @@ class Module {
     */
     void setRfSwitchPins(RADIOLIB_PIN_TYPE rxEn, RADIOLIB_PIN_TYPE txEn);
 
     /*!
-      \brief Set RF switch state.
+      \brief Some modules contain external RF switch controlled by pins.
+      This function gives RadioLib control over those pins to
+      automatically switch between various modes.
+
+      Radiolib will automatically set the pin mode and value of these
+      pins, so do not control them from the sketch.
+
+
+      \param pins A reference to an array of pins to control. This
+      should always be an array of 3 elements. If you need less pins,
+      use RADIOLIB_NC for the unused elements.
+
+      \param table A reference to an array of pin values to use for each
+      supported mode. Each element is an RfSwitchMode_T struct that
+      lists the mode for which it applies and the values for each of the
+      pins passed in the pins argument respectively.
+
+      The `pins` array will be copied into the Module object, so the
+      original array can be deallocated after this call. However,
+      a reference to the `table` array will be stored, so that array
+      must remain valid as long RadioLib is being used.
+
+      The `mode` field in each table row should normally use any of the
+      `MODE_*` constants from the Module::OpMode_t enum. However, some
+      radios support additional modes and will define their own OpMode_t
+      enum.
 
-      \param rxPinState Pin state to set on Tx enable pin (usually high to transmit).
+      The length of the table is variable (to support radios that add
+      additional modes), so the table must always be terminated with the
+      special END_OF_MODE_TABLE value.
 
-      \param txPinState  Pin state to set on Rx enable pin (usually high to receive).
+      Normally all modes should be listed in the table, but for some
+      radios, modes can be omitted to indicate they are not supported
+      (e.g. when a radio has a high power and low power TX mode but
+      external circuitry only supports low power). If applicable, this
+      is documented in the radio class itself.
+
+      #### Example
+      For example, on a board that has an RF switch with an enable pin
+      connected to PA0 and a TX/RX select pin connected to PA1:
+
+      \code
+      // In global scope, define the pin array and mode table
+      static const RADIOLIB_PIN_TYPE rfswitch_pins[] =
+                             {PA0,  PA1,  RADIOLIB_NC};
+      static const Module::RfSwitchMode_t rfswitch_table[] = {
+        {Module::MODE_IDLE,  {LOW,  LOW}},
+        {Module::MODE_RX,    {HIGH, LOW}},
+        {Module::MODE_TX,    {HIGH, HIGH}},
+         Module::END_OF_MODE_TABLE,
+      };
+
+      void setup() {
+        ...
+        // Then somewhere in setup, pass them to radiolib
+        radio.setRfSwitchTable(rfswitch_pins, rfswitch_table);
+        ...
+      }
+      \endcode
+
+      TODO: Should we put this 3 (also in other places) into a constant?
```

Comment:
The danger I see with defining a macro or constant (say `RFSWITCH_NUM_PINS=3`) is that sketches might start using this, e.g.:

```
      // In global scope, define the pin array and mode table
      static const RADIOLIB_PIN_TYPE rfswitch_pins[RFSWITCH_NUM_PINS] =
                             {PA0,  PA1,  RADIOLIB_NC};
      static const Module::RfSwitchMode_t rfswitch_table[] = {
        {Module::MODE_IDLE,  {LOW,  LOW}},
        {Module::MODE_RX,    {HIGH, LOW}},
        {Module::MODE_TX,    {HIGH, HIGH}},
         Module::END_OF_MODE_TABLE,
      };
```

Then, if we later decide to change the value to 4, the above code will define an extra pin element, which is initialized to 0 (or maybe even undefined if it is a local variable and not a global, but I'm not 100% sure about this) and will thus cause RadioLib to unexpectedly write values to pin 0.

For clarity, the upgrade path I had in mind when we ever *do* need to support 4 pins, is to increase the size of `_rfSwitchPins` and then add an additional overload of `setRfSwtichTable`, so you will have two overloads:

    void setRfSwitchTable(const RADIOLIB_PIN_TYPE (&pins)[3], const RfSwitchMode_t table[]);
    void setRfSwitchTable(const RADIOLIB_PIN_TYPE (&pins)[4], const RfSwitchMode_t table[]);

The former can ensure that the unused element is set to `NC`, so existing sketches that define 3-element pin arrays will just work as before. But this only works when the array size in sketches does not increase, hence my hesitancy of adding a constant for this.

(note that in this case the `RfSwitchMode_t` type does need to change from 3 pins to 4 pins, enlarging all existing `rfswitch_table` variables in sketches automatically, but extra zeroes do not hurt there. This means that no compatibility handling is needed for the tables).


Maybe adding a good compromise would be to add a `MAX_RFSWITCH_PINS` constant, adding "MAX" to make it less likely that people will use it in sketches, and maybe add some comments to explicitly discourage this? I don't think we can make this a protected constant in the Module class, because it will be part of the API (in `RfSwitchMode_t` and probably also in the `setRfSwitchTable` argument type).

Also, you say "It would be better IMO to define the number of pins as a macro", does that mean you prefer using a macro over a `const`? I usually prefer `const`, since those adhere to normal scoping rules and give better compiler errors when names are accidentally reused elsewhere. Also, making this a `const` inside the module class might help a bit more to hide it from users even if it cannot be made protected.

Reply:
The compromise with naming the macro (or constant) MAX* and documenting it properly is a good solution. At the end of the day, most people will keep just using the 2-pin RF switches, which is handled by `Module::setRfSwitchPins`. If their RF setup needs more complex driving, then they are free to define their own switch table.

Reply:
Resolved in 4e26a8a9e0f73f029b3efa3a6301d06bd7d3f29a